### PR TITLE
Support @validates(field_obj)

### LINKED ISSUE #367: RFC: Improved API for registering field validator methods
## Status quo

Validator methods are registered by decorating them with `@validates('my_field_name')`.

``` python
from marshmallow import fields, Schema, ValidationError, validates

class ItemSchema(Schema):
    quantity = fields.Int()

    @validates('quantity')
    def validate_quantity(self, value):
        if value > 30:
            raise ValidationError('Quantity must not be greater than 30')
```
## Proposed API

Validator methods are registered by decorating them with `@my_field.validator`.

``` python
from marshmallow import fields, Schema, ValidationError

class ItemSchema(Schema):
    quantity = fields.Int()

    @quantity.validator
    def validate_quantity(self, value):
        if value > 30:
            raise ValidationError('Quantity must not be greater than 30')
```
## Benefits
- Less prone to user error (no magic strings).
- Removes redundant code. Since `Field#validator()` decorator would just append to a field's `validators` attribute, there would be no need for [this code](https://github.com/marshmallow-code/marshmallow/blob/bad7e56cd5602daf4eea31ce3cb49f07a5d7a9ba/marshmallow/schema.py#L762-L800). 
- Better introspection of fields. Again, since `Field#validator()` appends to `Field#validators`, code that introspects fields (e.g. apispec) can find all of a field's validators in one place.

### PR DESCRIPTION
**Pull request by taion**
Closes https://github.com/marshmallow-code/marshmallow/issues/367

### CODE REVIEW

File: tests/test_decorators.py [hunk 1]
```diff
@@ -316,11 +340,21 @@ def validate_bar(self, value):
         assert '"bar" field does not exist.' in str(excinfo)
 
     def test_precedence(self):
-        class Schema2(ValidatesSchema):
+        class Schema2Base(Schema):
+            foo = fields.Int()
+
+            @validates('foo')
```
Comment (taion) [inline tests/test_decorators.py:346]: This is an interesting/tricky edge case.

I can't actually use the original `ValidatesSchema` here, because `@validates(foo)` tries to validate the `foo` field declared on `ValidatesSchema`.

However, that field gets overridden in `Schema2`, and thus can't be found when trying to validate it.

Comment (sloria) [reply tests/test_decorators.py:346]: So if I'm understanding this correctly, `@validates('foo')` and `@validates(foo)` behave differently:

- `@validates(foo)` will always validate the `foo` field instance passed in
- `@validates('foo')` searches for `foo` according to the MRO.

So users would have to be careful when doing `@validates(foo)` on base schema classes because subclasses won't be able to override `foo`.

**Review by sloria — COMMENTED** [review]
I like the name changes and refactoring done in cc69995...b822cb2. Perhaps we should get those merged into `dev` in separate PR so we can review the `@validates(field_obj)` code in isolation.

**Comment by taion** [comment]
I've moved the internal refactoring code to https://github.com/marshmallow-code/marshmallow/pull/814.

File: tests/test_decorators.py (hunk 1 above)
Comment (taion) [reply tests/test_decorators.py:346]: yes. thinking about it some more, this is pretty weird DX-wise.

a different way to do this might be to bind the fields to their names in `SchemaMeta`.

that way, assuming the field has been bound to its name, then we can look up the name of the field from the field object, and not have to care about the MRO. it would also let us get rid of much of the extra code here around building the extra mapping.

the downside, then, is that this would limit the re-use of field objects. so for example, right now i believe this is okay:

```python
field = fields.Integer()

class FooSchema(Schema):
    foo = field

class BarSchema(Schema):
    bar = field
```

the straightforward way to do this (binding the field object to the assigned name in `SchemaMeta`) would prohibit the above, though it would still allow

```python
field = fields.Integer()

class FooSchema(Schema):
    foo = field

class BarSchema(Schema):
    foo = field
```

**Comment by taion** [comment]
Rebased.

File: tests/test_decorators.py (hunk 1 above)
Comment (taion) [reply tests/test_decorators.py:346]: @sloria Any thoughts on the above?

**Comment by sloria** [comment]
https://github.com/marshmallow-code/marshmallow/issues/367#issuecomment-530502790
